### ISSUE TITLE
feature request: warn about local variables declared but not set inside loops

### ISSUE THREAD

**Issue by bmillwood** [koalaman/shellcheck#3364]
#### For new checks and feature suggestions
- [x] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related
    - the keywords I tried: local unset uninitialised uninitialized

#### Here's a snippet or screenshot that shows a potential problem:

```sh
#!/usr/bin/env bash
function f() {
  for i in a b c; do
    local v
    [ "$i" == b ] && v=true
    echo "$i: $v"
  done
}
f
```

This code outputs:

```
a: 
b: true
c: true
```

I polled some people I know and they frequently guessed the last line incorrectly -- they might expect that after declaring `local v` they're guaranteed that `v` is unset, but in a loop that's not true.

- See https://github.com/NixOS/nixpkgs/pull/462319 for a bug I found in the wild that stemmed from believing that if you'd just done a `local` variable declaration, it must have an unset value. 

#### Here's what shellcheck currently says:

No issues detected!

#### Here's what I wanted to see:

A new warning that `v` is function-local, not loop-local, and since it's uninitialised at the declaration site it can retain values across loop iterations. Should recommend either that the variable is declared `local` outside the loop or that it's initialised at the declaration. For extra credit, don't produce a warning if the variable is clearly set after the declaration before it can possibly be used (i.e. `local v; v=x` is equivalent to `local v=x` and doesn't trigger the warning).

(I think it's very possible that this is one instance of a more general phenomenon of people not understanding how `local` works, and there's a better warning that would have caught the aforementioned bug. But wanted to start a conversation about it regardless.)

**Comment by nexushoratio** [comment]
Would folks be surprised if the loop was unrolled?

```shell
f() {
  local v
  v=true
  echo "$v"
  local v
  echo "$v"
}
```

**Comment by bmillwood** [comment]
> Would folks be surprised if the loop was unrolled?
> 
> f() {
>   local v
>   v=true
>   echo "$v"
>   local v
>   echo "$v"
> }

I think people who expect `local` to be loop-scoped would not (necessarily) expect this to be equivalent to the original program. (I think they'd imagine that all local variables are implicitly unset at the end of each loop or something).

**Comment by nexushoratio** [comment]
Shell is such strange beast.

I've been using it for 35 years.  But nothing overly complicated.

It is only when when I started running shellcheck and trying to understand some stuff that I learned and (starting to) understand the dynamic scoping.

I'm not sure it makes sense for the tool to try to identify every possible issue unless it has good support for data flow analysis.

---

What if it example was slightly more complicated?

```shell
g() {
  unset v
}

f() {
  for i in a b c; do
    local v
    [ "$i" == b ] && v=true
    echo "$i: $v"
    g
  done
}
```

Then the `local` becomes important.  But unless shellcheck knows how to look into the call of `g()`, and analyze all of the various twists and turns it might take, it has no way of knowing if it is necessary or not inside that loop.

**Comment by bmillwood** [comment]
My check as specified would warn on your `f` (regardless of what `g` does) and I think that would be a reasonable thing to do.

(FWIW commenting out the `local` line in your code doesn't change the behaviour of `f` as far as I can see.)

**Comment by wileyhy** [comment]
In Bash, IIRC, the only way to change the definition of any parameter,
alias or function is either to reassign some other value to that entity or
to use the 'unset' built-in.

  The man page is intended to be the authoritative technical document; a
reliable link to the current version is at the bottom of this email.

  It's interesting: this function 'g' could unset either a global or a
local variable depending on where it's called.

g() {
  unset v
}

...oh yes, and 'unalias'.


local [option] [name[=value] ... | - ]
  For each argument, create a local variable named name and assign it
value. The option can be any of the options accepted by declare. When local
is used within a function, it causes the variable name to have a visible
scope restricted to that function and its children. It is an error to use
local when not within a function.
  If name is -, it makes the set of shell options local to the function in
which local is invoked: any shell options changed using the set builtin
inside the function after the call to local are restored to their original
values when the function returns. The restore is performed as if a series
of set commands were executed to restore the values that were in place
before the function.
  With no operands, local writes a list of local variables to the standard
output.
  The return status is 0 unless local is used outside a function, an
invalid name is supplied, or name is a readonly variable.

https://tiswww.case.edu/php/chet/bash/bash.html

**Comment by nexushoratio** [comment]
> (FWIW commenting out the `local` line in your code doesn't change the behaviour of `f` as far as I can see.)

It changes whether `v` is local to `f()` or global to the script.

Checking at the top level with something like `declare -p ` or `echo ${v?}` would expose the differences.
